Settings page: Improve UI for post/page types to track

acicovic opened the pull request:
## Description
In the settings page, the UI for selecting the post/page types to track needs to be improved as to make it more straightforward and user friendly.

## Motivation and Context
- A more thorough discussion about this can be found in #483
- Closes #483
- Closes #643

## How Has This Been Tested?
- Manual testing of data saving and loading, including wiping the plugin's settings in the database to see how the change reacts to default/unsaved settings
- Testing that built CSS works by removing and re-enabling its wp_enqueue_style() function
- Checking on a browser that the table is responsive and works well in big and small widths
- Integration tests pass

## Screenshots
### Before:
![image](https://user-images.githubusercontent.com/23142906/152368215-6cea5406-f178-451f-952c-8792f6ed977d.png)

### After:
![image](https://user-images.githubusercontent.com/23142906/152038520-04e66bd9-9237-4aef-a2bd-ccefcffc1703.png)

acicovic committed:
Settings: Fix function description

acicovic committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

acicovic committed:
Settings page: Add new Track Post Type UI prototype

acicovic commented:
Hello there,

I've just committed the UI prototype for the Track Post Types functionality, the aim being to experience the UI as well as examine the code from an accessibility perspective.

Things to know:
- I've introduced a new CSS file to hold admin CSS, but I haven't created a built file for it (yet?).
- This is a UI-only commit and I haven't tested what happens when the form saves. A saving/loading commit will come later. For the time being, you're saving at your own risk!
- Fun fact: while trying to commit this, I stumbled upon #639

Regarding accessibility
- I've replaced the usual `<legend>` that we have in most fieldsets in the settings page with a `<caption>` within the table.
- Tabbing works as expected.
- In this case, accessibility can be tricky and there are many different screen readers. I decided to use `aria-labelledby`.

Please let me know of any findings, suggestions, etc.

acicovic committed:
Settings Page: Implement load/save for 'Track Post Types as'

acicovic committed:
Settings page: Add safety check during post tracking validation

acicovic committed:
Settings page validation: Check for missing required arrays only when needed

acicovic committed:
Remove role="presentation" from 'Track Post Types as...' table

acicovic commented:
Hi, I'm opening this for review.

SonarCloud is complaining because `class-settings-page.php` now has 21 methods from the "20 max allowed". It also suggests adding a `<legend>` in the `<fieldset>`, but I'm not sure this is needed in our case because we do have the `<caption>` within the `<table>`. If anyone has a better understanding of this, please let me know.

Thanks!

acicovic committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -285,41 +287,19 @@ function (): void {
-			'track_page_types',
-			__( 'Page Types to Track', 'wp-parsely' ),
-			array( $this, 'print_multiple_checkboxes' ),
+			__( 'Track Post Types as...', 'wp-parsely' ),
```
I would remove the `...` from the label, so leave it as `Track Post Types as`

acicovic replied:
Not pushed yet, but done.

acicovic commented:
Noting here some related improvements for completeness:
- ~~Build the CSS file~~ - Created as a new issue for all our CSS files in #643 
- ~~Add new UI screenshot to our WP.org page~~ - Added to #609 
 - [x] Make adjustments so the whole table is visible in very small screens (for example320px) when the post type name is really long

acicovic committed:
Remove '...' from 'Track post type as...'

acicovic committed:
Settings page: Adjust 'Track Post Types as' table for small widths

When a post type name was very long, the 'Track Post Types as' table would get out of screen bounds in really small widths (<370px).

acicovic committed:
Merge branch 'add/new-settings-post-type-tracking-ui' of https://github.com/Parsely/wp-parsely into add/new-settings-post-type-tracking-ui

acicovic committed:
Rename 'parsely_admin_settings_css' to 'wp-parsely-admin-style'

acicovic committed:
Settings page: Add built CSS file using webpack

acicovic committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

acicovic committed:
Remove unneeded blank line

acicovic committed:
Fix failing E2E tests

While performing E2E testing, the function changeKeysState() would result in a WordPress critical error which resulted in tests timing out. This would only happen on the testing environment as this wasn't reproducible on the development environment.

It would seem that under the test environment, some arrays weren't being set and this was making code inside the validate_options_post_type_tracking() function to crash, resulting in the WordPress critical error. I was never able to reproduce this in my development environment, but the fix is simple and allows everything to work correctly in all environments.

pauarge committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

acicovic committed:
Simplify validate_options_post_type_tracking() function

acicovic committed:
Merge branch 'add/new-settings-post-type-tracking-ui' of https://github.com/Parsely/wp-parsely into add/new-settings-post-type-tracking-ui

acicovic committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

acicovic committed:
print_track_post_types_table(): Add legend tag to obey to SonarCloud

acicovic committed:
Merge branch 'add/new-settings-post-type-tracking-ui' of https://github.com/Parsely/wp-parsely into add/new-settings-post-type-tracking-ui

pauarge committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -76,23 +76,34 @@ public function run(): void {
 			$admin_settings_asset = require plugin_dir_path( PARSELY_FILE ) . 'build/admin-settings.asset.php';
+			$built_assets_url     = plugin_dir_url( PARSELY_FILE ) . '/build/';
+
+			// Admin settings JS.
```
I think this comment and the one in line 100 are redundant.

acicovic replied:
Removed both.

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -442,41 +453,19 @@ function (): void {
 		);
 
 		// Allow use of custom taxonomy to populate articleSection in parselyPage; defaults to category.
-		$h          = __( 'By default, Parse.ly only tracks the default post type as a post page. If you want to track custom post types, select them here!', 'wp-parsely' );
```
Could we keep this help text (or something that's similar enough but fits with the current UI)?

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -442,41 +453,19 @@ function (): void {
-			'label_for'      => Parsely::OPTIONS_KEY . "[$field_id]",
-		);
+		$field_id   = 'track_post_types_as';
+		$field_help = __( 'Select how you want to track every Post Type.', 'wp-parsely' );
```
No need to capitalize `Post Type`

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -653,7 +642,7 @@ public function print_description_text( $args ) {
 
 	/**
-	 * Print out the radio buttons.
+	 * Print out an input text tag.
```
👍

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -821,6 +810,89 @@ public function print_media_single_image( array $args ) {
 
+	/**
+	 * Print out the post tracking options table.
+	 *
```
We're missing a `@since` annotation here.

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -821,6 +810,89 @@ public function print_media_single_image( array $args ) {
+	/**
+	 * Print out the post tracking options table.
+	 *
+	 * @param array $args The arguments used in the output HTML elements.
```
Would it be possible to make the `array` type more concrete here, something similar to what we do in: https://github.com/Parsely/wp-parsely/blob/89c057b5783f99b90d9d7a37659b53221eee8503/src/Integrations/class-amp.php#L158?

(if the object type is too complex to represent, feel free to dismiss my comment)

acicovic replied:
So far we're only passing string key-value pairs so I've changed it to `array<string, string>`. By the way, syntax highlighting in VSCode for this isn't super great.

pauarge replied:
Thanks for the change. This really helps PHPStan to infer types, in my setup (PHPStorm) works pretty well.

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -821,6 +810,89 @@ public function print_media_single_image( array $args ) {
+	 *
+	 * @return array Key-value pairs with post type and their 'track as' value.
+	 */
+	public function get_tracking_values_for_display() {
```
Missing return type

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -1054,6 +1089,63 @@ public function validate_options( array $input ): array {
+	 */
+	private function validate_options_post_type_tracking( array &$input ): void {
+		$options         = $this->parsely->get_options();
+		$posts           = 'track_post_types';
+		$pages           = 'track_page_types';
```
Maybe we could pull these two out of the function as constants?

acicovic replied:
I thought of this myself, but in the bigger picture I would consider this as part of #564 as an `options` class or similar which would expose constants for every option key. However, I can do this as an intermediary step if you wish.

pauarge replied:
Fair, we can leave it as is.

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -1094,7 +1186,7 @@ private static function get_logo_default(): string {
 	 * @return array
 	 */
-	private static function validate_option_array( array $array ): array {
+	private static function sanitize_option_array( array $array ): array {
```
Why the name change?

acicovic replied:
This was done for clarity, as this function doesn't validate, but rather _**sanitizes**_ the array passed to it. As such, `sanitize` is much more descriptive of what the function does in reality, and makes it stand out from the functions starting with `validate`.

As this is a private function, changing it wasn't risky, so I went ahead with it.

pauarge reviewed (changes requested):
Amazing work! Just some small tweaks left.

acicovic committed:
Settings page: Adjust comments to satisfy code review

acicovic committed:
Merge branch 'add/new-settings-post-type-tracking-ui' of https://github.com/Parsely/wp-parsely into add/new-settings-post-type-tracking-ui

acicovic committed:
get_tracking_values_for_display(): Add return type

acicovic committed:
Settings page: Fix help text capitalization

acicovic committed:
Settings page: Harmonize help text with existing help texts

pauarge committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -452,7 +452,7 @@ function (): void {
 		// Allow use of custom taxonomy to populate articleSection in parselyPage; defaults to category.
 		$field_id   = 'track_post_types_as';
-		$field_help = __( 'Select how you want to track every post type.', 'wp-parsely' );
+		$field_help = __( 'By default, Parse.ly only tracks posts and pages. If you want to track other post types, select how you want to track them here!', 'wp-parsely' );
```
Can we remove the final exclamation mark? The text looks good otherwise.

acicovic committed:
Settings page: Remove exclamation mark

acicovic committed:
Merge branch 'add/new-settings-post-type-tracking-ui' of https://github.com/Parsely/wp-parsely into add/new-settings-post-type-tracking-ui

acicovic committed:
Settings page: Don't allow saving tracking settings for inexistent post types

acicovic committed:
Don't allow saving tracking settings with invalid values

acicovic committed:
Settings page: Remove unneeded duplicate tracking detection

acicovic committed:
Settings page: Update saving tracking values test

acicovic committed:
Add E2E tests for "Track Post Types as" setting

acicovic committed:
Merge branch 'develop' into add/new-settings-post-type-tracking-ui

pauarge commented on tests/e2e/utils.js:
```diff
@@ -45,3 +45,43 @@ export const checkH2DoesNotExist = async ( text ) => {
+/**
+ * Save settings in the settings page and force a hard refresh.
+ */
+export const saveSettingsAndHardRefresh = async () => {
```
Could we have similar docstring as the method above in this one?

pauarge reviewed (changes requested):
Super small change, nice work on the testing!

acicovic committed:
utils.js: Fix comment for saveSettingsAndHardRefresh()

pauarge commented on src/UI/class-settings-page.php:
```diff
@@ -76,23 +76,32 @@ public function run(): void {
 			wp_enqueue_script(
-				'parsely_admin_settings',
-				plugin_dir_url( PARSELY_FILE ) . '/build/admin-settings.js',
+				'parsely-admin-settings-js',
```
Could we remove the `js` suffix to this script and the `css` to the CSS one? They'd have the same name but be different by the fact that one is a `script` enqueued and the other is a `style`.

acicovic committed:
Remove js/css suffixes for admin setting asset handles

pauarge reviewed (dismissed):
LGTM

acicovic committed:
Use 'non-existent' instead of 'inexistent' (suggested by Gary Jones)
